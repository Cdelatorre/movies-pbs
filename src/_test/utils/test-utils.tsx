import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { AppStore, RootState, setupStore } from "../../store";
import { BrowserRouter, Router } from "react-router-dom";
import { emptyStoreData } from "./mocks/mock";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = emptyStoreData,

    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
  route?: string
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (route) window.history.pushState({}, "Test routing", route);

    return (
      <BrowserRouter>
        <Provider store={store}>{children}</Provider>;
      </BrowserRouter>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithRouter(ui: React.ReactElement, history?: any) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (history) {
      return (
        <Router location="/" navigator={history}>
          {children}
        </Router>
      );
    } else {
      return <BrowserRouter>{children}</BrowserRouter>;
    }
  }
  return { ...render(ui, { wrapper: Wrapper }) };
}
