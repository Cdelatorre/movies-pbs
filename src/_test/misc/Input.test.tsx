import { fireEvent, screen } from "@testing-library/react";
import { renderWithProviders } from "../utils/test-utils";
import Input from "../../components/misc/input";

describe("Renders Input component correctly", () => {
  test("Renders Input default component correctly", () => {
    renderWithProviders(<Input name="test-name" label="Label test" />);

    const input = screen.getByTestId("input-text-test-name");
    expect(input).toBeInTheDocument();
    expect(screen.getByText("Label test")).toBeInTheDocument();

    fireEvent.change(input, {
      target: { value: "Movie mock title" },
    });
    expect(screen.getByDisplayValue("Movie mock title")).toBeInTheDocument();
  });

  test("Input is modified it's default behavior by adding props and fires specific events", () => {
    const onChange = jest.fn();
    const onBlur = jest.fn();
    const onFocus = jest.fn();
    const onClick = jest.fn();

    renderWithProviders(
      <Input
        name="test-name"
        label="Label test"
        type="number"
        value="1"
        error="test message error"
        onChange={onChange}
        onBlur={onBlur}
        onFocus={onFocus}
        onClick={onClick}
      />
    );

    const input = screen.getByTestId("input-number-test-name");
    expect(input).toBeInTheDocument();

    expect(screen.getByDisplayValue("1")).toBeInTheDocument();
    expect(screen.getByText("Label test")).toBeInTheDocument();
    expect(screen.getByText("test message error")).toBeInTheDocument();
    expect(input).toHaveClass("form-number border-danger");

    fireEvent.change(input, { target: { value: "Tests value" } });
    fireEvent.click(input);
    fireEvent.focus(input);
    fireEvent.blur(input);

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onBlur).toHaveBeenCalledTimes(1);
    expect(onFocus).toHaveBeenCalledTimes(1);
    expect(onClick).toHaveBeenCalledTimes(1);
  });
});
