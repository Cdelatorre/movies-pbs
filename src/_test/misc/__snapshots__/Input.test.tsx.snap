// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Input component correctly Renders Input default component correctly 1`] = `
<input
  class="form-control form-text bg-dark text-light"
  data-testid="input-text-test-name"
  name="test-name"
  type="text"
  value=""
/>
`;
